client_unix.go

```diff
@@ -5,15 +5,71 @@ package containerd
 import (
 	"fmt"
 	"net"
+	"os"
 	"strings"
+	"syscall"
 	"time"
+
+	"github.com/pkg/errors"
 )
 
-func dialer(address string, timeout time.Duration) (net.Conn, error) {
+func isNoent(err error) bool {
+	if err != nil {
+		if nerr, ok := err.(*net.OpError); ok {
+			if serr, ok := nerr.Err.(*os.SyscallError); ok {
+				if serr.Err == syscall.ENOENT {
+					return true
+				}
+			}
+		}
+	}
+	return false
+}
+
+type dialResult struct {
+	c   net.Conn
+	err error
+}
+
+func Dialer(address string, timeout time.Duration) (net.Conn, error) {
+	var (
+		dr    *dialResult
```

Comment:
don't make this a func level var and just return it from the select when you get it back on the chan